crowd_anki/export/anki_exporter.py

```diff
@@ -67,3 +71,81 @@ def _copy_media(self, deck, deck_directory):
                             str(media_directory.resolve()))
             except IOError as ioerror:
                 print("Failed to copy a file {}. Full error: {}".format(file_src, ioerror))
+
+    def sort_notes(self, notes):
+        deck_sort_config = self.config.get(EXPORT_DECK_SORT, False)
+
+        if not deck_sort_config or "method" not in deck_sort_config:
+            return notes
+
+        sort_method = deck_sort_config["method"]
+        is_reversed = deck_sort_config["reversed"] if "reversed" in deck_sort_config else False
+
+        def _formatted_sort_method(method):
+            return re.sub("[-_\s+]", "", method).lower()
+
+        # Make sure formatting of sort_method is correct, and format each entry
+        if isinstance(sort_method, str):
+            sort_method = [_formatted_sort_method(sort_method)]
+        elif isinstance(sort_method, list):
+            if len(sort_method) == 1:
+                sort_method = [_formatted_sort_method(sort_method[0])]
+            elif len(sort_method) > 1:
+                sort_method = [_formatted_sort_method(sm) for sm in sort_method]                
+            else:
+                return notes
+        else:
+            return notes
+
+        # Create mapping of sortable_values variable, to sort_method keys, to sortable_values lambda
+        sorting_definitions = [
+            (0, ("guid", "guids"), lambda i: i.anki_object.guid),
+            (1, ("flag", "flags"), lambda i: i.anki_object.flags),
+            (2, ("tag", "tags"), lambda i: i.anki_object.tags),
+            (3, ("notemodel", "notemodels", "notemodelname", "notemodelsname", "notemodelnames", "notemodelsnames"), lambda i: i.anki_object._model["name"]),
+            (4, ("notemodelid", "notemodelsid", "notemodelids", "notemodelsids"), lambda i: i.anki_object._model["crowdanki_uuid"]),
+            (5, ("field1", "firstfield", "first"), lambda i: i.anki_object.fields[0]),
+            (6, ("field2", "secondfield", "second"), lambda i: i.anki_object.fields[1])
+        ]
+
+        # Extract each notes values which could be used for sorting into sortable_values, for use later on
+        for note in notes:
+            note.sortable_values = [None] * len(sorting_definitions)
+            note.export_filter_set.add("sortable_values")
+
+            for defs in sorting_definitions:
+                note.sortable_values[defs[0]] = defs[2](note)
```

Comment:
if I understand this correctly you apply all key functions in advance here? why do it here and not inline?

also, please refrain from modifying the notes, if you really need to inject additional information for sorting - look into creating a list of pairs, where you sortable information would be in one of the parts of the pair, and note would be the other. Then you'd sort that, extract notes and replace the notes list in the deck (but really I think you should just apply keys inline)

Reply:
I saw that you had lots of other ignored values on the notes and decided to go with your convention. A paired list is a good idea. Where exactly do you mean when you say "inline"? As in, when the notes are created? If so, the simple answer is that since this code was all kept within this function I had no way (or intention) of making the values available to other parts of the code. Now that I am going to separate it into it's own class that should be doable though. I'll make a sortconfig file which will contain the definitions, which can be accessed in other parts if need be.

Reply:
From the phone on the plane now, but will try to answer in the time I have :p. 
Another way to put this is why not create a lamda that would apply multiple functions to value and return a tuple, instead of just returning values precomputed here?

Reply:
so to give you an example of what I'm talking about, I'd create a sort key function like:

```
# assuming a simple name -> key map

keys = (sorting_definitions[method_name] for method_name in sort_method)
def sort_key(note):   
    return (key(note) for key in keys)

notes = sorted(notes, key=sort_key...)
```

Reply:
>I saw that you had lots of other ignored values on the notes and decided to go with your convention. 

I'm not really fond of those, they are mainly there as a way to filter out the object variables, but I don't think I inject the new values dynamically anywhere..

Reply:
This was very helpful, thank you. Returning the value inline is a good idea, rather than my ham-fisted approach!

I was trying to do Tuple comprehension before to accomplish this, just like how you have given as an example, but it would never work like I expected. Your suggestion here got me to try again, and after some digging I found that `return (key(note) for key in keys)` will not return a Tuple of keys, but instead a Generator expression. You need to be explicit instead with `return tuple(key(note) for key in keys)`. So my google-fu found :sweat_smile: https://stackoverflow.com/questions/16940293/why-is-there-no-tuple-comprehension-in-python

Anyways I have pushed a change, I hope the result is nicer to your Python sensibilities :smirk: (it is to mine!)